packages/uif/src/indexers/multi/MultiIndexer.ts

```diff
@@ -0,0 +1,208 @@
+   * `setStoredConfigurations` itself.
+   */
+  abstract multiInitialize(): Promise<SavedConfiguration<T>[]>
+
+  /**
+   * Implements the main data fetching process. It is up to the indexer to
+   * decide how much data to fetch. For example given `.update(100, 200, [...])`, the
+   * indexer can only fetch data up to 110 and return 110. The next time this
+   * method will be called with `.update(110, 200, [...])`.
+   *
+   * @param from The height for which the indexer should start syncing data.
+   * This value is inclusive. If the indexer hasn't synced anything previously
+   * this will equal the minimum height of all configurations.
+   *
+   * @param to The height at which the indexer should end syncing data. This
+   * value is also inclusive so the indexer should eventually sync data for this
+   * height.
+   *
+   * @param configurations The configurations that the indexer should use to
+   * sync data. The configurations are guaranteed to be in the range of
+   * `from` and `to`. Some of those configurations might have been synced
+   * previously for this range. Those configurations will include the `hasData`
+   * flag set to `true`.
+   *
+   * @returns The height that the indexer has synced up to. Returning
+   * `from` means that the indexer has synced a single data point. Returning
+   * a value greater than `from` means that the indexer has synced up
+   * to that height. Returning a value less than `from` or greater than
+   * `to` is not permitted.
+   */
+  abstract multiUpdate(
+    from: number,
+    to: number,
+    configurations: UpdateConfiguration<T>[],
+  ): Promise<number>
+
+  /**
+   * Removes data that was previously synced but because configurations changed
+   * is no longer valid. The data should be removed for the ranges specified
+   * in each configuration. It is possible for multiple ranges to share a
+   * configuration id!
+   *
+   * This method can only be called during the initialization of the indexer,
+   * after `multiInitialize` returns.
+   */
+  abstract removeData(configurations: RemovalConfiguration<T>[]): Promise<void>
+
+  /**
+   * Saves configurations that the indexer should use to sync data. The
+   * configurations saved here should be read in the `multiInitialize` method.
+   *
+   * @param configurations The configurations that the indexer should save. The
+   * indexer should save the returned configurations and ensure that no other
+   * configurations are persisted.
+   */
+  abstract saveConfigurations(
+    configurations: SavedConfiguration<T>[],
+  ): Promise<void>
+
+  async initialize(): Promise<number> {
+    const saved = await this.multiInitialize()
+    const { toRemove, toSave, safeHeight } = diffConfigurations(
+      this.configurations,
+      saved,
+    )
+    this.saved = toSave
+    if (toRemove.length > 0) {
+      await this.removeData(toRemove)
+    }
+    await this.saveConfigurations(toSave)
```

Comment:
if there is `toRemove.length > 0` check then maybe `toSave.length > 0` check should also be there?

Reply:
Because we want to save the info that we now have 0 configs